Size_t-ify libssl

mattcaswell opened the pull request:
This is a more-or-less complete conversion of the whole implementation of libssl to use size_t for sizes of things where appropriate. It is the second part of two PRs (see #1750 for the first part). A third PR will be required with a small amount of work to join the two halves together (although they can be reviewed and implemented entirely independently).

This introduces SSL_read_ex(), SSL_write_ex() and SSL_peek_ex() which use size_t arguments and return size_t information about the number of bytes read/written. All internal structure members and function arguments have been converted to use size_t where appropriate.

The boundary of this conversion is libssl. This means there are a number of "TODOs" left in the code where we call into libcrypto, and in those locations various type conversions need to take place.

I have not converted other libssl APIs which should probably have been size_t from the start, and instead do the necessary casts/conversions internally where needed. Over time we may want to convert some of those too.

mattcaswell commented:
This also should resolve all warnings in the libssl part of the compilation for the VC-WIN6A target on windows - all those warnings were almost exclusively about implicit conversions between size_t and int.

mattcaswell commented:
Travis was failing because there were some clashing function error codes introduced by an earlier rebase. New commit pushed to resolve it.

mattcaswell commented:
Travis was still failing with a bogus uninit variable warning, and also a shadowed variable declaration warning. New commits added.

dwmw2 commented:
How do you print `size_t` in tests, given some platforms like Windows haven't reached 1999 yet and don't support `%zu`? Cf. https://github.com/openssl/openssl/pull/1666#issuecomment-255157660

richsalz commented on crypto/hmac/hmac.c:
```diff
@@ -118,7 +118,10 @@ int HMAC_Final(HMAC_CTX *ctx, unsigned char *md, unsigned int *len)
 {
-    return EVP_MD_size((ctx)->md);
+    int size = EVP_MD_size((ctx)->md);
+    if (size < 0)
```
blank line after the decl and use the ?: operator.

richsalz commented on doc/ssl/SSL_CTX_set_mode.pod:
```diff
@@ -34,26 +34,27 @@ The following mode changes are available:
+when just a single record has been written). This works in a similar way for
+SSL_write(). When not set (the default), SSL_write_ex() or SSL_write() will only
+report success once the complete chunk was written. Once SSL_write_ex() or
+SSL_write() returns successful, r bytes have been written and the next call to
```
params in bold, so ```B<r>```

richsalz commented on doc/ssl/SSL_CTX_set_mode.pod:
```diff
@@ -34,26 +34,27 @@ The following mode changes are available:
+SSL_write(). When not set (the default), SSL_write_ex() or SSL_write() will only
+report success once the complete chunk was written. Once SSL_write_ex() or
+SSL_write() returns successful, r bytes have been written and the next call to
+SSL_write_ex() or SSL_write() must only send the n-r bytes left, imitating the
```
```B<n-r>```

richsalz commented on doc/ssl/SSL_CTX_set_split_send_fragment.pod:
```diff
@@ -58,19 +58,19 @@ Pipelining operates slightly differently for reading encrypted data compared to
+SSL_write(_ex) called with 4001-6000 bytes == 3 pipelines used
 
-SSL_write called with 6001+ bytes == 4 pipelines used
+SSL_write(_ex) called with 6001+ bytes == 4 pipelines used
```
avoid this terse syntax and do ```SSL_write()/SSL_write_ex()```

richsalz commented on doc/ssl/SSL_get_error.pod:
```diff
@@ -64,10 +64,11 @@ TLS/SSL I/O function should be retried.
-to read data.  This is mainly because TLS/SSL handshakes may occur at any
-time during the protocol (initiated by either the client or the server);
-SSL_read(), SSL_peek(), and SSL_write() will handle any pending handshakes.
+SSL_read_ex(), SSL_read() SSL_peek_ex() or SSL_peek() may want to write data and
```
missing commas after read,peek_ex, and peek.

richsalz commented on doc/ssl/SSL_get_error.pod:
```diff
@@ -64,10 +64,11 @@ TLS/SSL I/O function should be retried.
+SSL_write() or SSL_write_ex() may want to read data.  This is mainly because
+TLS/SSL handshakes may occur at any time during the protocol (initiated by
+either the client or the server); SSL_read_ex(), SSL_read(), SSL_peek_ex(),
+SSL_peek(), SSL_write_ex() and SSL_write() will handle any pending handshakes.
```
missing commas; a, b, c (oxford comma :)

richsalz commented on doc/ssl/SSL_read.pod:
```diff
@@ -2,82 +2,100 @@
+SSL_peek_ex() and SSL_peek() are identical to SSL_read_ex() and SSL_read()
+respectively except no bytes are actually removed from the underlying BIO during
+the read, so that a subsequent call to SSL_read_ex() or SSL_read() will yield
+the same bytes.
```
at least the same bytes?

richsalz commented on doc/ssl/SSL_read.pod:
```diff
@@ -2,82 +2,100 @@
 =head1 NOTES
 
-If necessary, SSL_read() will negotiate a TLS/SSL session, if
+In this notes section all comments that apply to SSL_read_ex() or SSL_read()
```
In the paragraphs below, ...

richsalz commented on doc/ssl/SSL_read.pod:
```diff
@@ -2,82 +2,100 @@
 
-When an SSL_read() operation has to be repeated because of
+When an SSL_read_ex() or SSL_read() operation has to be repeated because of
 B<SSL_ERROR_WANT_READ> or B<SSL_ERROR_WANT_WRITE>, it must be repeated
```
```SSL_get_error() returned``` ?

richsalz commented on doc/ssl/SSL_set_connect_state.pod:
```diff
@@ -35,8 +35,8 @@ requested, the handshake routines must be explicitly set.
 routines are automatically set. When performing a transparent negotiation
-using L<SSL_write(3)> or L<SSL_read(3)>, the
-handshake routines must be explicitly set in advance using either
+using L<SSL_write_ex(3)>, L<SSL_write(3)> L<SSL_read_ex(3)> or L<SSL_read(3)>,
```
missing commas.

richsalz commented on doc/ssl/SSL_read.pod:
```diff
@@ -2,82 +2,100 @@
-records (with a maximum record size of 16kB for SSLv3/TLSv1). Only when a
-record has been completely received, it can be processed (decryption and
+SSL_read_ex() and SSL_read() work based on the SSL/TLS records. The data are
+received in records (with a maximum record size of 16kB for SSLv3/TLSv1). Only
```
drop the "for ..." at least kill mention of SLSv3, which then makes it a truism so just drop the "for..." part.

richsalz commented on doc/ssl/SSL_read.pod:
```diff
@@ -2,82 +2,100 @@
-record has been completely received, it can be processed (decryption and
+SSL_read_ex() and SSL_read() work based on the SSL/TLS records. The data are
+received in records (with a maximum record size of 16kB for SSLv3/TLSv1). Only
+when a record has been completely received, it can be processed (decryption and
```
"it can" --> "can it"

richsalz commented on doc/ssl/SSL_read.pod:
```diff
@@ -2,82 +2,100 @@
-If the underlying BIO is B<blocking>, SSL_read() will only return, once the
-read operation has been finished or an error occurred, except when a
-renegotiation take place, in which case a SSL_ERROR_WANT_READ may occur.
+call of SSL_read_ex() or SSL_read() can still be buffered inside the SSL layer
```
perhaps "read call" instead of "call of read or read_ex" in some places?  and is it pointed out explicitly that read and read_ex can be intermixed?

richsalz commented on doc/ssl/SSL_read.pod:
```diff
@@ -2,82 +2,100 @@
+and will be retrieved on the next call to SSL_read_ex() or SSL_read(). If B<num>
+is higher than the number of bytes buffered, SSL_read_ex() or SSL_read() will
+return with the bytes buffered. If no more bytes are in the buffer, SSL_read()
+will trigger the processing of the next record. Only when the record has been
```
doesn't read_ex also trigger this?  hence my suggestion about "read call" usage.

mattcaswell commented:
New commits pushed. I think I got all your comments. I made similar changes to the SSL_write() page as you suggested for SSL_read().

levitte commented on ssl/bio_ssl.c:
```diff
@@ -101,14 +102,14 @@ static int ssl_read(BIO *b, char *out, int outl)
     switch (SSL_get_error(ssl, ret)) {
     case SSL_ERROR_NONE:
-        if (ret <= 0)
+        if (readbytes == 0)
```
Is that even possible?  `SSL_get_error()` will only return `SSL_ERROR_NONE` when `ret > 0`, and that would never happen when only zero bytes have been read, right?

levitte replied:
So are you telling me that `SSL_read_ex()` might return 1 when 0 bytes have been read?  I thought you defined that as a failure...  Now, observe [these lines](https://github.com/openssl/openssl/blob/master/ssl/ssl_lib.c#L2912-L2913) in `SSL_get_error()`...  it seems it will only return `SSL_ERROR_NONE` if `ret > 0`.

Of course, it doesn't hurt...  but it *is* confusing when why is trying to understand the return codes from `SSL_read_ex()`

mattcaswell replied:
I'm getting myself confused with too many PRs in this area at the moment (and the return values are really screwed up). #1763 actually changes things so that reading 0 from the BIO results in -1 being returned from SSL_read(), which would be 0 for SSL_read_ex(). So, on reflection, I think you're right. I'll remove this.

levitte commented on ssl/bio_ssl.c:
```diff
@@ -171,17 +173,14 @@ static int ssl_write(BIO *b, const char *out, int outl)
     switch (SSL_get_error(ssl, ret)) {
     case SSL_ERROR_NONE:
-        if (ret <= 0)
+        if (written == 0)
```
Same reasoning here as in `ssl_read()`

mattcaswell replied:
As above

kroeckx commented:
I guess I need to submit my review comments or they don't show up.

kroeckx commented on ssl/record/ssl3_record.c:
```diff
@@ -381,9 +389,9 @@ int ssl3_get_record(SSL *s)
     }
 #ifdef SSL_DEBUG
-    printf("dec %d\n", rr->length);
+    printf("dec %ld\n", rr->length);
```
%ld is not the right specifier for a size_t. %d is a signed type while size_t is unsigned, and a size_t doesn't have to have the same size as a long and doesn't at least on 64 bit windows. In C99 you should use %zu. We probably need something like PRIu32 for platforms that don't support %z.

kroeckx replied:
Probably.

dwmw2 replied:
[Allegedly](https://blogs.msdn.microsoft.com/vcblog/2014/06/03/visual-studio-14-ctp/#div-comment-77743) even Microsoft has finally caught up with the tail end of the 20th century.  How feasible is it to just use `%zu` from OpenSSL $NEXT?

mattcaswell commented on doc/ssl/SSL_read.pod:
```diff
@@ -2,82 +2,100 @@
 
-If necessary, SSL_read() will negotiate a TLS/SSL session, if
+In this notes section all comments that apply to SSL_read_ex() or SSL_read()
+also apply to SSL_peek_ex() and SSL_peek().
```
I already did in the previous commits.

mattcaswell commented on doc/ssl/SSL_read.pod:
```diff
@@ -2,82 +2,100 @@
+SSL_read_ex() will return 1 for success or 0 for failure. In the event of a
+failure call SSL_get_error() to find out the reason.
+
+For SSL_read() the following return values can occur:
```
Fixed.

mattcaswell commented on include/internal/constant_time_locl.h:
```diff
@@ -102,12 +103,22 @@ static ossl_inline unsigned int constant_time_msb(unsigned int a)
+{
+    return 0 - (a >> (sizeof(a) * 8 - 1));
+}
+
```
I added tests to constant_time_test for this.

mattcaswell commented on ssl/packet_locl.h:
```diff
@@ -189,6 +202,19 @@ __owur static ossl_inline int PACKET_get_net_3(PACKET *pkt, unsigned long *data)
     return 1;
 }
 
+/* Same as PACKET_get_net_3() but for a size_t */
```
Yes, its still used.

mattcaswell commented:
New commits pushed addressing review feedback.

kroeckx commented on ssl/bio_ssl.c:
```diff
@@ -93,6 +93,7 @@ static int ssl_read(BIO *b, char *out, int outl)
     SSL *ssl;
     int retry_reason = 0;
     int r = 0;
+    size_t read;
```
Because read is the name of a function, which is probably why you got that warning.

mattcaswell commented:
New commits pushed to remove the unneeded conditions from ssl_read()/ssl_write() in bio_ssl, and to document the HMAC_size() function.

kroeckx commented on doc/ssl/SSL_read.pod:
```diff
@@ -2,82 +2,95 @@
 
-The following return values can occur:
+SSL_read_ex() and SSL_peek_ex() will return 1 for success or 0 for failure. In
+the event of a failure call SSL_get_error() to find out the reason.
```
I'm wondering if we should describe the difference in error cases between the _ex() and non-ex() function. For instance is having no data available and retrying an error? And how do I know that it's retryable?

mattcaswell replied:
I added some more detail to this (and to the SSL_write() page).

richsalz commented:
Will start reading the rest in a couple of days; travel tomorrow.

richsalz commented:
Add a long cast?

levitte commented:
> Add a long cast?

@richsalz, I think you need to make those comments using the web interface.  By email, this became just a general comment unrelated to anything at all...

mattcaswell commented:
Clarified the documentation as per feedback. Also added a commit to fix a travis failure.

kaduk commented on doc/ssl/SSL_read.pod:
```diff
@@ -2,82 +2,100 @@
+Failures can be retryable (e.g. we are waiting for more bytes to
+be delivered by the network) or non-retryable (e.g. a fatal network error). In
+the event of a failure call L<SSL_get_error(3)> to find out the reason and if
+the call is retryable or not.
```
And there is a magic error reason that will be present to indicate a retryable call?  (What is it?)

mattcaswell replied:
All that is documented SSL_get_error() (which is referenced). Surely we don't need to repeat it here?

mattcaswell replied:
I added your alternative text (to this page and the SSL_write page)

levitte commented on include/openssl/e_os2.h:
```diff
@@ -276,6 +276,17 @@ typedef unsigned __int64 uint64_t;
 # endif
 
+/* Format specifier for printing size_t */
+# if (defined(__STDC_VERSION__) && __STDC_VERSION__ >= 199901L)
```
Unfortunately, you need to add an extra check :-(

``` C
# if (defined(__STDC_VERSION__) && __STDC_VERSION__ >= 199901L) && !defined(__VMS)
```

mattcaswell replied:
Fixed.

levitte replied:
Actually, ignore that.  I'll tell the compiler to use the C94 standard instead.

kroeckx replied:
I'm getting confused. VMS has a C99 compiler that doesn't support %zu, so you switch to C94 (which doesn't exist?) instead?

Anyway this is implemented by libc, not the compiler. So even if your compiler doesn't support %zu your libc might.

mattcaswell commented:
New commits pushed to fix VMS issue and hopefully resolve Travis problems once and for all!

levitte commented on include/openssl/e_os2.h:
```diff
@@ -277,7 +277,8 @@ typedef unsigned __int64 uint64_t;
 /* Format specifier for printing size_t */
-# if (defined(__STDC_VERSION__) && __STDC_VERSION__ >= 199901L)
+# if (defined(__STDC_VERSION__) && __STDC_VERSION__ >= 199901L) \
+      && !defined(__VMS)
```
With 4f3015bb30b7d95bb97408776b70e6a35fb91e8a, the VMS check isn't necessary any more.  I've checked.

kroeckx commented on include/openssl/e_os2.h:
```diff
@@ -276,6 +276,18 @@ typedef unsigned __int64 uint64_t;
+#  define OSSLzu  "zu"
+# else
+#  ifdef SIXTY_FOUR_BIT_LONG
+#   define OSSLzu PRIu64
```
I think SIXTY_FOUR_BIT_LONG does something other than you expect. Some 64 bit platforms define SIXTY_FOUR_BIT and not SIXTY_FOUR_BIT_LONG. I'm not sure defining either one of them is even required, but I guess most actually do.

But the PRIu64 thing is also a C99 thing, so this doesn't make any sense.

levitte replied:
Could it be that the `#else` part should look like this?

``` C
# else
#  ifdef THIRTY_TWO_BIT
#   define OSSLzu "lu"
#  else
#   define OSSLzu PRIu64
#  endif
# endif
```

Note that `THIRTY_TWO_BIT` is the default unless we say otherwise in the config targets.

kroeckx replied:
I think that both actually have a problem. In theory size_t could be an int and not a long, and so you'll get a warning that it's expecting a long and gets an int. (It could even be a short, but that gets promoted to an int anyway.) So we might need both "u" and "lu". I guess we should just wait until someone actually complains about this.

But the PRIu64 is probably more problematic, because we're on platform that doesn't have C99. So we could at least check that PRIu64 is a define or not. We should probably use "%llu" here, but I think I at least remember some platforms that wanted "%lu" even if long and long long are the same size.

levitte replied:
Wait, hold on.  *We* define `PRIu64` unconditionally if it isn't already defined, just a few lines up.  So the C99 comment isn't very relevant.

kroeckx replied:
I was looking in the wrong file ...

mattcaswell replied:
I came up with the same construction as @levitte...now in #1787

dwmw2 commented:
It would be good if we could land just the `OSSLzu` format string separately, to unblock #1666. I'm holding up the next OpenConnect release until I have a coherent story about the payload MTU calculation (I have it working for <= 1.1.0 but would like to know it's future-proof before shipping it).

mattcaswell commented:
> It would be good if we could land just the OSSLzu format string separately

I pulled it out into #1787

mattcaswell commented:
Removed the OSSLzu change from this and rebased to pick it up from master.

mattcaswell commented:
Ping? Any more comments on this?

kroeckx commented:
It seems to have merge conflicts now ...

kroeckx commented:
Anyway, I didn't have time to actually look at everything in this patch yet. But I'll try to find some time.

richsalz commented:
Is this ABI compatible?

mattcaswell commented:
> Is this ABI compatible?

Yes

mattcaswell commented:
> It seems to have merge conflicts now ...

I rebased to remove the merge conflicts.

richsalz commented:
i do not think this is compatible.  size_t can be unsigned,  existing code using int code blow up.

mattcaswell commented:
> i do not think this is compatible. size_t can be unsigned, existing code using int code blow up.

What specifically is not compatible? Give me an example. All existing APIs remain the same (using int). So externally there is no difference, except there are a small number of additional APIs available where you can use size_t if you want to. This is about the internals of libssl.

richsalz commented on ssl/packet_locl.h:
```diff
@@ -189,6 +202,19 @@ __owur static ossl_inline int PACKET_get_net_3(PACKET *pkt, unsigned long *data)
+    unsigned long i;
+    int ret;
+
+    ret = PACKET_get_net_3(pkt, &i);
```
should me merged up into the decl line, but not a show-stopper.

richsalz commented on ssl/packet_locl.h:
```diff
@@ -242,6 +281,19 @@ __owur static ossl_inline int PACKET_get_1(PACKET *pkt, unsigned int *data)
+    unsigned int i;
+    int ret;
+
+    ret = PACKET_get_1(pkt, &i);
```
merge?

mattcaswell replied:
Done. And for a couple of others like this.

richsalz commented on ssl/record/rec_layer_s3.c:
```diff
@@ -287,11 +288,14 @@ int ssl3_read_n(SSL *s, int n, int max, int extend, int clearold)
         if (max < n)
             max = n;
-        if (max > (int)(rb->len - rb->offset))
+        if (max > (rb->len - rb->offset))
```
remove the parens around the subtraction.

richsalz commented on ssl/record/rec_layer_s3.c:
```diff
@@ -593,14 +602,15 @@ int ssl3_write_bytes(SSL *s, int type, const void *buf_, int len)
         }
 
-        if ((i == (int)n) ||
+        if ((tmpwrit == n) ||
```
remove parens around the tmpwrit test.

mattcaswell replied:
Done

kaduk commented:
> size_t can be unsigned

Sorry for the pedanticism, but size_t is required to be unsigned in both C89 and C99.

stevelinsell commented on ssl/pqueue.c:
```diff
@@ -141,7 +141,7 @@ pitem *pqueue_next(pitem **item)
+size_t pqueue_size(pqueue *pq)
 {
     pitem *item = pq->items;
     int count = 0;
```
Any reason why the internal representation count stays as an int when the function interface has changed to a size_t? I assumed this would change internally as well?

mattcaswell replied:
Yes...because it was missed! ;-) Will fix

mattcaswell replied:
Fixed.

mattcaswell commented:
I rebased this, and fixed some minor merge conflicts due to the TLS1.3 stuff that has gone in so far. Also added a couple of commits to fix the style issues, and the issue noted by @stevelinsell.

@richsalz can you reconfirm?

richsalz commented:
+1 reconfirm.

mattcaswell commented:
Merged. Thanks!
